### PR TITLE
Add UInt160 and UInt256

### LINKED ISSUE #208: Add UInt160 and UInt256 data types
**Summary or problem description**

As UInt160, UInt256 are one of the most commonly used data types, I think we should abstract them  into basic types. Here is a [switcheo contract](https://github.com/Switcheo/switcheo-neo/blob/58bbb947d41f0bbec9563231b9f90f39f0f7bfbf/switcheo/BrokerContract.cs), as you can see, scripthash(UInt160) are widely used.

**Neo Version**
- Neo 3

**Where in the software does this update applies to?**
- VM
- Neo
- Compiler

### PR DESCRIPTION
**Pull request by ShawnYun**
Close #208

### CODE REVIEW

File: src/Neo.SmartContract.Framework/Helper.cs [hunk 1]
```diff
@@ -160,6 +160,84 @@ public static byte ToByte(this int source)
...
+            {
+                var str = hexString.Substring(i * 2, 2);
+                var byte1 = str.Substring(0, 1).StringToByte();
+                var byte2 = str.Substring(1, 1).StringToByte();
+                var byte3 = byte1 << 4;
+                var res = (byte)(byte3 + byte2);
+                result[i] = res;
+            }
+            return new UInt160(result.Reverse());
+        }
```
Comment (erikzhang) [inline src/Neo.SmartContract.Framework/Helper.cs:181]: I think this is too complicated for a contract and may cost a huge amount of GAS.

Comment (ShawnYun) [reply src/Neo.SmartContract.Framework/Helper.cs:181]: Or we can add a InteropService?

Comment (ShawnYun) [reply src/Neo.SmartContract.Framework/Helper.cs:181]: For static variables, we can do the conversion in the compiler. 
But for function arguments, we can't get the parameter values at compile time.

Comment (erikzhang) [reply src/Neo.SmartContract.Framework/Helper.cs:181]: Can you give a scenario?

Comment (ShawnYun) [reply src/Neo.SmartContract.Framework/Helper.cs:181]: For example:
```
static readonly string NeoToken = "0xde5f57d430d3dece511cf975a8d37848cb9e0525";
```
we can get the "0xde5f57d430d3dece511cf975a8d37848cb9e0525" in Compiler.
But if we have a function:
```
        public static UInt256 TestUInt256(String str)
        {
            UInt256 result = str.HexStringToUInt256();
            return result;
        }
```
the Function arguments are passed in when the contract is called.
This is after compiling，so compiler can't handle it.

Comment (erikzhang) [reply src/Neo.SmartContract.Framework/Helper.cs:181]: But why do you use a `string` as the parameter? I think you should use `UInt256` directly.

Comment (ShawnYun) [reply src/Neo.SmartContract.Framework/Helper.cs:181]: On the other hand, the conversion is not used too often, so the GAS consumption is acceptable.

Comment (erikzhang) [reply src/Neo.SmartContract.Framework/Helper.cs:181]: > We can't use a UInt256 as a contract parameter.

Why?

Comment (ShawnYun) [reply src/Neo.SmartContract.Framework/Helper.cs:181]: You mean use `public static UInt256 TestUInt256(UInt256 str)` 
and we transfer a `String` as a parameter?

Comment (devhawk) [reply src/Neo.SmartContract.Framework/Helper.cs:181]: > We can't use a `UInt256` as a contract parameter.

Doesn't Neo already support using UInt160/256 as a contract parameter type? https://github.com/neo-project/neo/blob/master/src/neo/SmartContract/ContractParameterType.cs#L12

**Comment by devhawk** [comment]
What's the progress on this? Can we get this in for Preview 4 https://github.com/neo-project/neo/issues/1936?

File: tests/Neo.SmartContract.Framework.UnitTests/TestClasses/Contract_UInt.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Neo.SmartContract.Framework;
+
+namespace Neo.Compiler.MSIL.TestClasses
+{
+    public class Contract_UInt : SmartContract.Framework.SmartContract
+    {
+        public static UInt160 TestUInt160(String str)
```
Comment (devhawk) [inline tests/Neo.SmartContract.Framework.UnitTests/TestClasses/Contract_UInt.cs:10]: We should have tests that take a UInt160/256 as an input parameter too

**Comment by devhawk** [comment]
We should also add ECPoint support to smart contract framework. It behaves similarly to UInt160/256. However, we might want to do this as a separate PR

File: src/Neo.Compiler.MSIL/FuncExport.cs [hunk 3]
```diff
@@ -41,7 +41,9 @@ internal static string ConvType(TypeReference t)
                 case "System.Int64":
                 case "System.UInt64":
                 case "System.Numerics.BigInteger": return "Integer";
-                case "System.Byte[]": return "ByteArray";
+                case "System.Byte[]":
+                case "Neo.SmartContract.Framework.UInt160":
```
Comment (devhawk) [inline src/Neo.Compiler.MSIL/FuncExport.cs:45]: Why are we returning UInt160/256 as a byte array?

File: src/Neo.SmartContract.Framework/NonemitWithConvertAttribute.cs [hunk 4]
```diff
@@ -7,6 +7,7 @@ public enum ConvertMethod
         HexToBytes,
         ToScriptHash,
         ToBigInteger,
+        ToUInt160,
```
Comment (shargon) [inline src/Neo.SmartContract.Framework/NonemitWithConvertAttribute.cs:10]: ToUInt256?

**Review by devhawk — CHANGES_REQUESTED** [review]
Obsoleted by https://github.com/neo-project/neo-devpack-dotnet/pull/362

**Comment by ShawnYun** [comment]
As #362, Performing conversion through VM is expensive, so we can close it.

**Pull request closed without merging**